lib/sass/script/functions.rb

```diff
@@ -2078,17 +2092,51 @@ def map_get(map, key)
     # @example
     #   map-merge(("foo": 1), ("bar": 2)) => ("foo": 1, "bar": 2)
     #   map-merge(("foo": 1, "bar": 2), ("bar": 3)) => ("foo": 1, "bar": 3)
+    #   map-merge(("foo": 1, "bar": 2), bar, 3) => ("foo": 1, "bar": 3)
+    #   map-merge((a: (b: c)), a, b, d) => (a: (b: d))
     # @overload map_merge($map1, $map2)
     #   @param $map1 [Sass::Script::Value::Map]
     #   @param $map2 [Sass::Script::Value::Map]
     # @return [Sass::Script::Value::Map]
     # @raise [ArgumentError] if either parameter is not a map
-    def map_merge(map1, map2)
+    # @overload map_merge($map1, $keys..., $map2)
+    #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value]]
+    #   @param $map2 [Sass::Script::Value::Map]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if either parameter is not a map
+    def map_merge(map1, *keys)
       assert_type map1, :Map, :map1
+      map2 = keys.pop
+      if map1.is_a? Sass::Script::Value::List
+        map1 = map(map1.to_h)
+      end
```

Comment:
So, coming back to this, the original implementation was done iteratively and only in the function itself. Suffice it to say that dealing with the frozen nature of maps and the multi-directional modifications that needed to happen, it was incredibly memory intensive. It was only when it dawned on me that I could attempt a recursive version, that anything approximating acceptably efficient code appeared.

Iterative algorithms are almost always more efficient than their recursive counterparts, but in this case, quickly "diving in" through the data structure, and only making the required changes, seems a better representation of what's happening here, than a traditional "recursive calculation". It's the fact that these are nested maps (like a graph) that make it reasonable to be recursive.

For now, I'm leaving it, but I cleaned up the API to simplify the above code. If you *really* want me to attempt an iterative version again, I'll do it, but otherwise I'm leaving it.